### ISSUE TITLE
Carthage support

### ISSUE THREAD

**Issue by angelolloqui** [pinterest/PINRemoteImage#25]
It would be nice to have carthage support. Not sure about the exact requirements, but I believe it needs to have a framework target.

**Comment by garrettmoon** [comment]
I'd love to add Carthage support, if you're interested in writing a pull request, I'd happily accept :)

**Comment by dcaunt** [comment]
See [Supporting Carthage for your framework](https://github.com/Carthage/Carthage#supporting-carthage-for-your-framework)

This involves creating an Xcode project with framework targets and sharing the schemes. Having the project file will generally benefit those not using CocoaPods. I may have a stab at this if I get time.

**Comment by pmcatominey** [comment]
Looking into this, it appears we'll need FLAnimatedImage to be available through Carthage first.

I've started playing with this in my fork.

**Comment by dcaunt** [comment]
I have `PINRemoteImage.framework` building locally with Carthage but I don't think it can work for Carthage users until FLAnimatedImage merges in https://github.com/Flipboard/FLAnimatedImage/pull/79.

In the meantime I can PR the framework support with FLAnimatedImage as a submodule which would solve https://github.com/pinterest/PINRemoteImage/issues/30.

**Comment by garrettmoon** [comment]
@dcaunt Awesome! Do you mean #22 ? It'd be great for the project if FLAnimatedImage could be optional!

**Comment by dcaunt** [comment]
Sorry, I'll try to explain this more clearly.

There are three issues here: 
#22 FLAnimatedImage is a non-optional dependency
#25 Carthage support (this issue)
#30 Framework target

Carthage support depends on #30 – we can't support building via Carthage without a framework. In terms of building a PINRemoteImage.framework, we need to add an xcodeproj, and link to FLAnimatedImage somehow. As FLAnimatedImage doesn't support Carthage yet, the best way is to include it as a git submodule and then the framework can be built in Xcode.

In terms of making FLAnimatedImage an optional dependency, I don't know how to achieve this off the top of my head for both CocoaPods and frameworks.

**Comment by pmcatominey** [comment]
@dcaunt Split FLAnimatedImage support into a separate repo, setup to depend on PINRemoteImage for both CocoaPods and Carthage?

**Comment by dcaunt** [comment]
That makes sense, I was looking for a way to keep everything together in one repo but I think a separate repo would be more straightforward.

**Comment by garrettmoon** [comment]
Hmm, let's see if we can't get FLAnimatedImage to support Carthage. I'll try and contact the author to get https://github.com/Flipboard/FLAnimatedImage/pull/79 merged in.

**Comment by exherb** [comment]
Flipboard/FLAnimatedImage#79  has been merged.

**Comment by markcleonard** [comment]
I see the Carthage Compatible label in the README, but in my attempt to use this with carthage I received:
`*** Skipped building PINRemoteImage due to the error:`
`Dependency "PINRemoteImage" has no shared framework schemes for any of the platforms: iOS`

**Comment by garrettmoon** [comment]
This should have been added by https://github.com/pinterest/PINRemoteImage/pull/77. I haven't done much Carthage, I'll try and see if I can't reproduce.

**Comment by bdorfman** [comment]
So if you don't specify a sha, branch, or tag in your Cartfile, Carthage will by default use the latest release according to Github's releases page. Since there hasn't been a new release since Carthage support was merged, this then fails. You should be able to fix by specifying "master" or any sha since Carthage support was merged (or @garrettmoon you could make a new release).

**Comment by garrettmoon** [comment]
Ah! Thanks @bdorfman I'll also make a new release :)

**Comment by bdorfman** [comment]
Also apparently if you have more than one Xcode project and none of them are in the root, the one it chooses to try to build frameworks for is actually undefined. The current setup works for me, but it is possible that it is dependent on the version of Carthage you are using or your OS or something. Possibly you should also move the Carthage xcodeproj to the root (currently in the PINRemoteImage directory) to ensure that it gets found instead of the test/example app.

**Comment by garrettmoon** [comment]
@markcleonard can you let us know if the above change worked?

**Comment by markcleonard** [comment]
*** Checking out Alamofire at "3.1.2"
*** Checking out SwiftyUserDefaults at "2.0.0"
*** Checking out PSOperations at "2.0.2"
*** Downloading PINRemoteImage.framework binary at "Carthage Support"
*** Downloading PINCache.framework binary at "Xcode 7 support"
*** Checking out RateLimit at "v1.2.1"
*** Checking out SwiftyJSON at "2.3.1"


Doesn't seem to check out in the same manner as the rest of my dependencies. It also doesn't build a scheme after.

**Comment by bdorfman** [comment]
That is totally fine. PINCache and PINRemoteImage have prebuilt binaries uploaded to github so Carthage will use those instead of building.

Although because FLAnimatedImage doesnt have a working Carthage yet, the prebuilt binary for PINRemoteImage doesn't support FLAnimatedImage I think.

**Comment by markcleonard** [comment]
@bdorfman @garrettmoon  Thanks! So in this way does it not recognize specifying iOS as a platform `carthage build --platform iOS`?  PINCache downloads to both Carthage/Build/iOS and Carthage/Build/Mac, whereas PINRemoteImage only downloads to Carthage/Build/iOS.

**Comment by dillan** [comment]
When I build using Carthage PINRemoteImage does not get built for iOS or Mac, and I'm only getting it to build for tvOS after adding some missing files in #173.

@bdorfman is there something missing that was accidentally not committed in #77? @garrettmoon eludes to that possibility here: https://github.com/pinterest/PINRemoteImage/pull/77#issuecomment-153418225

**Comment by mobileexpert1** [comment]
A shell task (/usr/bin/xcrun xcodebuild -workspace /Users/mobile/Desktop/Carthage/Checkouts/PINRemoteImage/Example-Mac/PINRemoteImage.xcworkspace -scheme PINRemoteImage -configuration Release CODE_SIGNING_REQUIRED=NO CODE_SIGN_IDENTITY= CARTHAGE=YES -showBuildSettings) failed with exit code 66:
xcodebuild: error: Scheme PINRemoteImage is not currently configured for the build action.

**Comment by nickvelloff** [comment]
```
A shell task (/usr/bin/xcrun xcodebuild -workspace /Carthage/Checkouts/PINRemoteImage/Example-Mac/PINRemoteImage.xcworkspace -scheme PINRemoteImage -configuration Release CODE_SIGNING_REQUIRED=NO CODE_SIGN_IDENTITY= CARTHAGE=YES -showBuildSettings) failed with exit code 66:
xcodebuild: error: Scheme PINRemoteImage is not currently configured for the build action.
```

**Comment by garrettmoon** [comment]
@mobileexpert1 @nickvelloff it's possible that something is configured incorrectly (I don't know much about Carthage), but that error happens when you try to build carthage from the root directory. It appears to work if you cd into the PINRemoteImage folder at the root of the project.

**Comment by nickvelloff** [comment]
@garrettmoon This would likely work, as it would remove the examples from the scope of the build. The problem is that when using this as a dependency in another project, I don't know how you would target a specific sub-directory to build. Reading around Carthage docs, I'm not sure that is possible.

**Comment by garrettmoon** [comment]
@nickvelloff well I'm happy to accept any PRs to improve the situation! Sadly I don't use Carthage so I'm not the best person to work on this :(

**Comment by nickvelloff** [comment]
@garrettmoon Right on. I'll whip one up.

**Comment by nickvelloff** [comment]
@garrettmoon Actually ASDK, which was the focus of this issue for me, has added `PINRemoteImage` to the official dependencies (optionally). This resolves the issue for me. I'm working with the team there to help with Carthage so PRs may be coming, but I want to see how things play out so not going to jump in quite yet. 
Thanks for this lib though!

**Comment by garrettmoon** [comment]
Sounds great! I work with those guys and gals so I think we can figure out how to get any PRs that come from it merged :)

**Comment by garrettmoon** [comment]
Hey, Carthage support got added, going to close this out.
